### ISSUE TITLE
Android build failing with newer Android SDK versions

### ISSUE THREAD

**Issue by Odd0002** [supertuxkart/stk-code#2853]
According to https://stackoverflow.com/questions/42912824/the-ant-folder-is-suddenly-missing-from-android-sdk-did-google-remove-it, the ant folder (and therefore build.xml files) used in android/build.xml to build SuperTuxKart for Android is not available in newer Android SDKs, resulting in build errors (building the APK) when trying to build STK for Android.

**Comment by deveee** [comment]
Indeed building with ant has been deprecated recently. We can always use older build tools - STK needs API 19 (or API 21 for aarch64 build). 

Tbh. I was never using anything other than ant for making apks and I see that at least SDL2 uses ant with build.xml too. I will try to find a solution that allows to use current SDK version.

Btw. Do you have any other issues with building the apk? I can try to improve the build scripts and documentation to make it easier for other people.

**Comment by Odd0002** [comment]
We could always include the default build.xml from ant (which I uploaded here: [buildxml.zip](https://github.com/supertuxkart/stk-code/files/1058616/buildxml.zip)) in the android folder and use that, unless there's licensing issues with that. I haven't tried building a release version yet, but other than this I was able to mostly just use the scripts provided, once I set up the SDK and NDK.

One small bug I noticed was that when generating the apk, ant failed because the SDK dir does not have another folder called "tools" in it. The SDK itself is in a folder called "tools", but build.xml says:  import file="${sdk.dir}/tools/ant/build.xml", which results in it trying to use "tools/tools/ant/build.xml". (If we include the build.xml in the android folder, this won't be a problem anymore.)

I also noticed that the generate_assets.sh doesn't use the variable SOUND_SAMPLE when resampling sounds and music.

**Comment by deveee** [comment]
I made a small fix for SOUND_SAMPLE variable.

We can't just include the build.xml because it also uses some .jar files (at least ant-tasks.jar).

Other possibilities are to use gradle or maven for creating the apk. But they both don't work out of the box on ubuntu (gradle has some issues with missing symbols and maven has android plugin in unsupported version...). Google seem suggest to use gradle for creating the apk from command line.

**Comment by deveee** [comment]
For building with gradle we will need something like this.
[build.gradle.txt](https://github.com/supertuxkart/stk-code/files/1066150/build.gradle.txt)

Just ANDROID_HOME and ANDROID_NDK_HOME must be set, SDK version changed to 19, set a path to Android.mk for ndk-build etc... And then apk can be made with ```gradle assembleDebug```

**Comment by deveee** [comment]
@Odd0002 Could you check if build with gradle works for you? Here is the patch:
```
diff --git a/android/build.gradle b/android/build.gradle
new file mode 100644
index 0000000000..c1b35396d3
--- /dev/null
+++ b/android/build.gradle
@@ -0,0 +1,31 @@
+buildscript
+{
+    repositories
+    {
+        jcenter()
+    }
+
+    dependencies
+    {
+        classpath 'com.android.tools.build:gradle:2.2.3'
+    }
+}
+
+apply plugin: 'com.android.application'
+
+android
+{
+    compileSdkVersion sdkVersion.toInteger()
+    buildToolsVersion "25.0.3"
+
+    sourceSets
+    {
+        main
+        {
+            manifest.srcFile 'AndroidManifest.xml'
+            jniLibs.srcDirs = ['libs']
+            res.srcDirs = ['res']
+            assets.srcDirs = ['assets']
+        }
+    }
+}
diff --git a/android/make.sh b/android/make.sh
index 331073a75e..278858aab9 100755
--- a/android/make.sh
+++ b/android/make.sh
@@ -22,16 +22,19 @@ export NDK_ABI_ARMV7=armeabi-v7a
 export ARCH_ARMV7=arm
 export HOST_ARMV7=arm-linux-androideabi
 export NDK_PLATFORM_ARMV7=android-19
+export SDK_VERSION_ARMV7=19
 
 export NDK_ABI_X86=x86
 export ARCH_X86=x86
 export HOST_X86=i686-linux-android
 export NDK_PLATFORM_X86=android-19
+export SDK_VERSION_X86=19
 
 export NDK_ABI_AARCH64=arm64-v8a
 export ARCH_AARCH64=arm64
 export HOST_AARCH64=aarch64-linux-android
 export NDK_PLATFORM_AARCH64=android-21
+export SDK_VERSION_AARCH64=21
 
 
 # A helper function that checks if error ocurred
@@ -75,16 +78,19 @@ if [ "$COMPILE_ARCH" = "armv7" ]; then
     export NDK_ABI=$NDK_ABI_ARMV7
     export ARCH=$ARCH_ARMV7
     export HOST=$HOST_ARMV7
+    export SDK_VERSION=$SDK_VERSION_ARMV7
 elif [ "$COMPILE_ARCH" = "x86" ]; then
     export NDK_PLATFORM=$NDK_PLATFORM_X86
     export NDK_ABI=$NDK_ABI_X86
     export ARCH=$ARCH_X86
     export HOST=$HOST_X86
+    export SDK_VERSION=$SDK_VERSION_X86
 elif [ "$COMPILE_ARCH" = "aarch64" ]; then
     export NDK_PLATFORM=$NDK_PLATFORM_AARCH64
     export NDK_ABI=$NDK_ABI_AARCH64
     export ARCH=$ARCH_AARCH64
     export HOST=$HOST_AARCH64
+    export SDK_VERSION=$SDK_VERSION_AARCH64
 else
     echo "Unknow COMPILE_ARCH: $COMPILE_ARCH. Possible values are: " \
          "armv7, aarch64, x86"
@@ -302,5 +308,16 @@ check_error
 
 # Build apk
 echo "Building APK"
-ant $BUILD_TYPE -Dsdk.dir="$SDK_PATH" -Dtarget=$NDK_PLATFORM
+#ant $BUILD_TYPE -Dsdk.dir="$SDK_PATH" -Dtarget=$NDK_PLATFORM
+
+export ANDROID_HOME="$SDK_PATH"
+
+if [ $IS_DEBUG_BUILD -gt 0 ]; then
+    GRADLE_BUILD_TYPE="assembleDebug"
+else
+    GRADLE_BUILD_TYPE="assembleRelease"
+fi
+
+gradle -PsdkVersion=$SDK_VERSION $GRADLE_BUILD_TYPE
+
 check_error
```
Atm. I hardcoded build-tools version to 25.0.3 because it's not able to automatically detect the latest available version. It requires version >= 19.1 for armv7 compilation.

**Comment by Odd0002** [comment]
The gradle build worked for both debug and release versions (through the make.sh file) and generated APKs for each, so it seems to be successful.
(However, I had some trouble with the licenses because I was linking the "tools" directory as the SDK, not the directory above it that contains the "tools" directory, so it wasn't finding the "licenses" directory and so wasn't letting me build it.)

I also got these warnings when building, but this isn't related to the patch provided above:
```
To run dex in process, the Gradle daemon needs a larger heap.
It currently has 1024 MB.
For faster builds, increase the maximum heap size for the Gradle daemon to at least 1536 MB.
To do this set org.gradle.jvmargs=-Xmx1536M in the project gradle.properties.
For more information see https://docs.gradle.org/current/userguide/build_environment.html
```

```
The JavaCompile.setDependencyCacheDir() method has been deprecated and is scheduled to be removed in Gradle 4.0.
The TaskInputs.source(Object) method has been deprecated and is scheduled to be removed in Gradle 4.0. Please use TaskInputs.file(Object).skipWhenEmpty() instead.
```

**Comment by deveee** [comment]
Ok, I changed the default build to gradle:
https://github.com/supertuxkart/stk-code/commit/ae19ade5e5d439c6a18f06affddb489d804e2a87
though there is still an option to use ant because I made much more tests with it and additionally it produces ~1-2MB smaller apk (better zip compression? no idea...). And it's actually simpler, just needs older SDK.

Regarding to the second warning, probably we will need to use higher android plugin version for gradle 4.0, but it's not compatible with gradle < 3.3, so we can change it later.
